feat(linear_algebra): Matrix inverses for square nonsingular matrices

Vierkantor opened the pull request:
This is intended as a first step towards defining the generic matrix inverse function. The function `nonsing_inv` applied to a square matrix `A` gives its inverse, as long as `A.det` is invertible. The more general pseudoinverse, which is also defined for singular / non-square matrices, should be defined in terms of this `nonsing_inv`, but that will be another pull request. I also propose that `has_inv (matrix n n R)` should be used for the pseudoinverse (but both should take the same value if they are (well-)defined).

The computation of the inverse is done via Cramer's rule, which is definitely not the most optimized way. It should still be computable in a reasonable time, and the correctness proof of Cramer's rule is not too long.

TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [x] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

Vierkantor committed:
Prove that some matrices have inverses

Vierkantor committed:
Finish the proof: show that the determinant is 0 if a column is repeated

Vierkantor committed:
Show that nonsingular_inv is also a right inverse

Vierkantor committed:
Cleanup and code movement

Vierkantor committed:
Small lemmata on transpose

Vierkantor committed:
WIP: some work on inverse matrices

Vierkantor committed:
Code cleanup and documentation

Vierkantor committed:
More cleanup and documentation

Vierkantor committed:
Generalize det_zero_of_column_eq to remove the linear order assumption

Vierkantor committed:
Fix linting issues.

Vierkantor committed:
Unneeded import can be removed

Vierkantor committed:
A little bit more cleanup

Vierkantor committed:
Generalize nonsing_inv to any ring with inverse

kbuzzard commented on src/linear_algebra/determinant.lean:
```diff
@@ -103,4 +104,122 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
+  so it doesn't work in characteristic `2`.
+-/
+lemma det_zero_of_column_eq_of_char_ne_two (char_ne_2 : ∀ (a : R), a = -a → a = 0) : M.det = 0 :=
+begin
```
I lecture this proof to the undergraduates because it's simpler than the general case -- but do we need two proofs of this in mathlib, one strictly less general than the other?

jcommelin replied:
Also: the condition `(char_ne_2 : ∀ (a : R), a = -a → a = 0)` is stronger than `char ≠ 2`. For example, in the ring `Z[X]/(2X)` you'll have `X = - X`, but `X ≠ 0`. Mathlib knows what the characteristic of a ring is: `char_p`. So if you actually want `char ≠ 2`, I guess you could assume `not (char_p R 2)`.

Vierkantor replied:
> do we need two proofs of this in mathlib, one strictly less general than the other?

True, this was really just an artifact of development that I shouldn't have left in the PR. I've removed the lemma now.

Vierkantor committed:
Improve comments for `nonsingular_inverse`

Vierkantor committed:
Remove the less general `det_zero_of_column_eq_of_char_ne_two` proof

ChrisHughes24 commented on src/algebra/big_operators.lean:
```diff
@@ -421,6 +421,22 @@ lemma prod_eq_one {f : α → β} {s : finset α} (h : ∀x∈s, f x = 1) : s.pr
+lemma prod_cancels_of_partition_cancels [R : setoid α] [decidable_rel R.r]
+  (h : ∀ x ∈ s, (s.filter (λy, y ≈ x)).prod f = 1) : s.prod f = 1 :=
+begin
+suffices : (s.image quotient.mk).prod (λ xbar, (s.filter (λ y, ⟦y⟧ = xbar)).prod f) = s.prod f,
```
tactics mode proofs should be indented.

Vierkantor replied:
I went through and indented all the tactical proofs, thanks!

ChrisHughes24 commented on src/group_theory/perm/sign.lean:
```diff
@@ -161,6 +161,15 @@ by rw [swap_mul_eq_mul_swap, inv_apply_self, inv_apply_self]
 equiv.swap_swap i j
 
+/-- A specialization of `swap_mul_self`, useful in cancelling out multiple swaps. -/
+lemma swap_mul_self_mul (i j : α) (σ : perm α) : equiv.swap i j * (equiv.swap i j * σ) = σ :=
```
this should `@[simp]`

ChrisHughes24 commented on src/group_theory/perm/sign.lean:
```diff
@@ -161,6 +161,15 @@ by rw [swap_mul_eq_mul_swap, inv_apply_self, inv_apply_self]
+lemma swap_mul_self_mul (i j : α) (σ : perm α) : equiv.swap i j * (equiv.swap i j * σ) = σ :=
+by rw [←mul_assoc (swap i j) (swap i j) σ, equiv.perm.swap_mul_self, one_mul]
+
+lemma swap_mul_eq_iff {i j : α} (σ : perm α) : swap i j * σ = σ ↔ i = j :=
```
I would make `σ` implicit here

ChrisHughes24 commented on src/linear_algebra/determinant.lean:
```diff
@@ -103,4 +104,100 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
+/-- Transposing a matrix preserves the determinant. -/
+@[simp] lemma det_transpose (M : matrix n n R) : M.transpose.det = M.det :=
+begin
+apply sum_bij (λ σ _, σ⁻¹),
```
tactics proofs should be indented again.

ChrisHughes24 commented on src/linear_algebra/determinant.lean:
```diff
@@ -103,4 +104,100 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
+end
+
+/-- Permuting the columns changes the sign of the determinant. -/
+lemma det_permute (M : matrix n n R) (σ : perm n) : matrix.det (λ i, M (σ i)) = σ.sign * M.det :=
```
I think I would prefer the convention of permuting the columns of a matrix by multiplying by a permutation matrix.

alexjbest replied:
Why not keep both versions though?

Vierkantor replied:
Aha, I thought you wanted the lemma statement to be replaced with a permutation matrix. Either lemma is a one-liner if you have the other, so I've just added the original `det_permute` back in as a new lemma.

ChrisHughes24 commented on src/linear_algebra/determinant.lean:
```diff
@@ -103,4 +104,100 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
+  We use this to partition permutations in the expression for the determinant,
+  such that each partitions sums up to `0`.
+-/
+def mod_swap {n : Type u} [decidable_eq n] (i j : n) : setoid (perm n) :=
```
This is the same as quotienting by the subgroup generated by `swap i j`. This is probably a nicer statement.

Vierkantor replied:
I tried rewording the statement, but unfortunately it seems like the original version is easier to work with.

The argument I want to use is that the determinant is a sum adding up to `0` because the term corresponding to `σ` cancels with the term corresponding to `swap i j * σ`. Thus, I want to partition the permutations so that the classes are `{σ, swap i j * σ}`, and then the sum over this class is `x + -1 * x + 0 = 0`.

Using the quotient of `group.closure {swap i j}` would then require as a first step showing that the classes are `{σ, swap i j * σ}`, which was immediate from the original `mod_swap` definition. (Is there a better way to show `quotient_group.left_rel (group.closure {swap i j}) σ τ ↔ (σ = τ ∨ σ = swap i j * τ)` than first using `group.closure {swap i j} = gpowers (swap i j) = {σ // ∃ k, σ = (swap i j)^k}`, then showing `(∃ k, σ = (swap i j)^k) ↔ (∃ 0 ≤ k < order_of (swap i j), σ = (swap i j)^k) ↔ (σ = id ∨ σ = swap i j)`?) So I don't expect the change is really worth the extra work.

ChrisHughes24 commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,313 @@
+
+/--
+  The `cramer_map` sends a matrix `A` and vector `b` to the vector `x`,
+  such that `A ⬝x = b`.
```
```suggestion
  such that `A ⬝ x = b`.
```

ChrisHughes24 commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,313 @@
+  such that `A ⬝x = b`.
+-/
+def cramer_map (i : n) : α := (A.replace_column i b)ᵀ.det
+lemma cramer_map_val (i : n) : cramer_map A b i = (A.replace_column i b)ᵀ.det := rfl
```
```suggestion
lemma cramer_map_def (i : n) : cramer_map A b i = (A.replace_column i b)ᵀ.det := rfl
```

ChrisHughes24 commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,313 @@
+  Depending on whether working coordinatewise is useful,
+  either version of Cramer's rule could be applied.
+-/
+def cramer_at (i : n) : (n → α) →ₗ[α] α :=
```
There's not much point defining `cramer_map` and this version. I would just keep the bundled version.

Vierkantor replied:
I decided to delete `cramer_at` and replace all references to `cramer_map` with `cramer` after the latter is defined. That should clear it up.

ChrisHughes24 reviewed (commented):
Some comments on the PR. I haven't gone through the whole thing yet.

Vierkantor committed:
Rename `cramer_map_val` -> `cramer_map_def`

Co-Authored-By: Chris Hughes <33847686+ChrisHughes24@users.noreply.github.com>

Vierkantor committed:
whitespace

Co-Authored-By: Chris Hughes <33847686+ChrisHughes24@users.noreply.github.com>

Vierkantor committed:
whitespace: indent tactic proofs

Vierkantor committed:
More renaming `cramer_map_val` -> `cramer_map_def`

Vierkantor committed:
`swap_mul_self_mul` can be a `simp` lemma

Vierkantor committed:
Make parameter σ to `swap_mul_eq_iff` implicit

Vierkantor committed:
Update usage of `function.update_same` and `function.update_noteq`

Vierkantor committed:
Replace `det_permute` with `det_permutation`

Although the statement now gives the determinant of a permutation matrix,
the proof is easier if we write it as a permuted identity matrix.
Thus the proof is basically the same, except a different line showing that the
entries are the same.

Vierkantor committed:
Re-introduce `matrix.det_permute` (now based on `matrix.det_permutation`)

Vierkantor committed:
Delete `cramer_at` and clean up the proofs depending on it

Vierkantor committed:
Replace `cramer_map` with `cramer` after defining `cramer`

ChrisHughes24 commented on src/linear_algebra/determinant.lean:
```diff
@@ -103,4 +106,101 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
+    { intros i _, simp },
+    { intros i j _ _ h, simp at h, assumption },
+    { intros i _, use σ⁻¹ i, finish }
+  },
```
Close brackets on same line usually.

sgouezel replied:
move to the previous line

ChrisHughes24 commented on src/linear_algebra/determinant.lean:
```diff
@@ -103,4 +106,101 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
+(λ h, or.inl h.symm)
+(λ h, or.inr (by rw [h, swap_mul_self_mul]))
+, λ σ τ υ hστ hτυ, by cases hστ; cases hτυ; try {rw [hστ, hτυ, swap_mul_self_mul]}; finish
+⟩
```
Could this be indented to make it obvious where the proof of each part starts.

sgouezel commented on src/data/matrix/pequiv.lean:
```diff
@@ -76,6 +76,10 @@ begin
 
+lemma to_pequiv_mul_matrix [semiring α] (f : n ≃ n) (M : matrix n n α) :
+  (f.to_pequiv.to_matrix * M) = λ i, M (f i) :=
+  by { ext i j, rw [mul_eq_mul, mul_matrix_apply, equiv.to_pequiv_apply] }
```
```suggestion
by { ext i j, rw [mul_eq_mul, mul_matrix_apply, equiv.to_pequiv_apply] }
```

sgouezel commented on src/data/matrix/basic.lean:
```diff
@@ -317,6 +329,9 @@ begin
 end
 
+@[simp] lemma transpose_smul [comm_ring α] (c : α)(M : matrix m n α) :
+  (c • M)ᵀ = c • Mᵀ := by { ext i j, refl }
```
```suggestion
  (c • M)ᵀ = c • Mᵀ := 
by { ext i j, refl }
```

sgouezel commented on src/data/matrix/pequiv.lean:
```diff
@@ -128,4 +132,8 @@ by rw [← to_matrix_trans, single_trans_single_of_ne hb, to_matrix_bot]
 
+/-- We can also define permutation matrices by permuting the rows of the identity matrix. -/
+lemma equiv_to_pequiv_to_matrix [has_one α] [has_zero α] (σ : equiv n n) (i j : n) :
+  σ.to_pequiv.to_matrix i j = (1 : matrix n n α) (σ i) j := if_congr option.some_inj rfl rfl
```
```suggestion
  σ.to_pequiv.to_matrix i j = (1 : matrix n n α) (σ i) j :=
if_congr option.some_inj rfl rfl
```

sgouezel commented on src/linear_algebra/determinant.lean:
```diff
@@ -1,10 +1,12 @@
 import data.matrix.basic
+import data.matrix.pequiv
 import group_theory.perm.sign
+import tactic.find
```
Remove?

sgouezel commented on src/linear_algebra/determinant.lean:
```diff
@@ -103,4 +106,101 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
+    congr,
+    { norm_num },
+    { ext i, apply pequiv.equiv_to_pequiv_to_matrix }
+  },
```
move to the previous line

sgouezel commented on src/linear_algebra/determinant.lean:
```diff
@@ -103,4 +106,101 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
+-/
+def mod_swap {n : Type u} [decidable_eq n] (i j : n) : setoid (perm n) :=
+⟨ λ σ τ, σ = τ ∨ σ = swap i j * τ
+, λ σ, or.inl (refl σ)
```
formatting in this proof with commas at the beginning of lines looks weird.

sgouezel commented on src/linear_algebra/determinant.lean:
```diff
@@ -103,4 +106,101 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
+
+instance (i j : n) : decidable_rel (mod_swap i j).r := λ σ τ, or.decidable
+
+variables {M : matrix n n R} {i j : n} (i_ne_j : i ≠ j) (hij : M i = M j)
```
could you include the Prop variables `i_ne_j` and `hij` in the statements of the lemmas instead of the variable list, to help the readers understand the statements?

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,306 @@
+lemma replace_column_ne {i' : n} (i_ne : i' ≠ i) : replace_column A i b i' = A i' :=
+function.update_noteq i_ne b A
+lemma replace_row_ne {j' : n} (j_ne : j' ≠ j) : replace_row A j b i j' = A i j' :=
+function.update_noteq j_ne (b i) (A i)
```
All these lemmas could be simp lemmas.

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,306 @@
+end replace
+
+section cramer
+/--
```
```suggestion
/-!
```

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,306 @@
+-/
+def cramer_map (i : n) : α := (A.replace_column i b).det
+
+lemma cramer_map_is_linear (i : n) : is_linear_map α (λ b, cramer_map A b i) := begin
```
```suggestion
lemma cramer_map_is_linear (i : n) : is_linear_map α (λ b, cramer_map A b i) :=
begin
```

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,306 @@
+    erw [finset.filter_eq', if_pos (mem_univ i), prod_singleton, prod_singleton,
+      prod_singleton, ←add_mul],
+    refl
+  },
```
Move to previous line

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,306 @@
+    erw [finset.filter_eq', if_pos (mem_univ i),
+      prod_singleton, prod_singleton, mul_assoc],
+    refl
+  }
```
move to previous line

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,306 @@
+    { apply prod_eq_zero (mem_univ j),
+      rw [replace_row_self],
+      exact if_neg h
+    },
```
move

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,306 @@
+    apply if_neg,
+    intro h',
+    exact h ((symm_apply_eq σ).mp h'.symm)
+  }
```
move

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,306 @@
+  rw [pi.smul_apply, smul_eq_mul, mul_ite]
+end
+
+lemma mul_adjugate (A : matrix n n α) : A ⬝adjugate A = A.det • 1 := begin
```
```suggestion
lemma mul_adjugate (A : matrix n n α) : A ⬝ adjugate A = A.det • 1 := begin
```

sgouezel reviewed (commented):
mostly cosmetic comments

Vierkantor committed:
Apply suggestions from code review

Co-Authored-By: sgouezel <sebastien.gouezel@univ-rennes1.fr>

Vierkantor committed:
Clean up imports

Vierkantor committed:
Formatting: move } to previous lines

Vierkantor committed:
Move assumptions of `det_zero_of_repeated_column` from variable to argument

sgouezel commented on src/algebra/big_operators.lean:
```diff
@@ -222,14 +222,23 @@ calc s.prod (λ x, h (if p x then f x else g x))
   simp only [filter_eq, if_true, if_false, h, prod_empty, prod_singleton, insert_empty_eq_singleton],
 end
 
+/-- Similar to `prod_ite_eq` but with arguments to `eq` swapped. -/
```
A good docstring should be understandable just by itself (when it appears in a tooltip in vscode), so better explain what the lemma says, and then mention `prod_ite_eq`.

Vierkantor replied:
I don't know how to write a docstring for this lemma that doesn't restate the type in words. Perhaps I should just delete the docstring here?

digama0 replied:
There is nothing wrong with a doc string that restates the type in words. It improves greppability simply by using the english terms informally.

sgouezel commented on src/data/finset.lean:
```diff
@@ -727,6 +727,11 @@ begin
     rintros m ⟨e⟩, exact h m, }
 end
 
+/-- A variant of `filter_eq` with the equality the other way. -/
```
Clarify docstring

sgouezel commented on src/data/finset.lean:
```diff
@@ -727,6 +727,11 @@ begin
 
+/-- A variant of `filter_eq` with the equality the other way. -/
+lemma filter_eq' [decidable_eq β] (s : finset β) (b : β) :
+  s.filter(λ a, a = b) = ite (b ∈ s) {b} ∅ :=
```
```suggestion
  s.filter (λ a, a = b) = ite (b ∈ s) {b} ∅ :=
```

Vierkantor replied:
Shall I also make this change in `filter_eq`?
```lean
  s.filter(eq b) = ite (b ∈ s) {b} ∅ :=
```
->
```lean
  s.filter (eq b) = ite (b ∈ s) {b} ∅ :=
```

digama0 replied:
yes, function application always gets a space

sgouezel commented on src/group_theory/perm/sign.lean:
```diff
@@ -161,6 +161,16 @@ by rw [swap_mul_eq_mul_swap, inv_apply_self, inv_apply_self]
 @[simp] lemma swap_mul_self (i j : α) : equiv.swap i j * equiv.swap i j = 1 :=
 equiv.swap_swap i j
 
+/-- A specialization of `swap_mul_self`, useful in cancelling out multiple swaps. -/
```
Improve docstring

sgouezel commented on src/algebra/big_operators.lean:
```diff
@@ -421,6 +430,22 @@ lemma prod_eq_one {f : α → β} {s : finset α} (h : ∀x∈s, f x = 1) : s.pr
 
+/-- If we can partition a product into subsets that cancel out, then the whole product cancels. -/
+@[to_additive]
+lemma prod_cancels_of_partition_cancels [R : setoid α] [decidable_rel R.r]
```
In your application, `R` is a rather ad hoc setoid instead of a completely canonical setoid that comes as a type class. And I expect this will be the case most of the time. Then maybe you should use `R` as a regular parameter instead of a typeclass parameter (but register it as a typeclass with `letI` in the proof).

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,300 @@
+open_locale matrix
+open equiv equiv.perm finset
+
+section replace
```
Since you are using `function.update`, why not go with `update_column` and so on for coherence?

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,300 @@
+/--
+  The `cramer_map` sends a matrix `A` and vector `b` to the vector `x`,
+  such that `A ⬝ x = b`.
+-/
```
I don't see how the docstring relates to the definition (in general, if `A` is not onto, there is no `x` with `Ax=b`...)

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,300 @@
+need to define those first. The file also doesn't define a `has_inv` instance
+for `matrix` so that can be used for the pseudoinverse instead.
+
+The definition of inverse used in this file is the one given by Cramer's rule.
```
It would help to give the definition of `cramer` and `adjugate` in the docstring, as these are not completely standard notions.

ChrisHughes24 commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,300 @@
+@[simp] lemma replace_column_self : replace_column A i b i = b := function.update_same i b A
+@[simp] lemma replace_row_self : replace_row A j b i j = b i := function.update_same j (b i) (A i)
+
+@[simp] lemma replace_column_eq {i' : n} (h : i = i') : replace_column A i b i' = b :=
```
This seems like a slightly strange way to state the lemma. Why not `@[simp] lemma replace_column_eq : replace_column A i b i = b`?

Vierkantor replied:
That lemma already existed, so I just replaced usages of `replace_column_eq` with `replace_column_self` and an extra rewriting step.

ChrisHughes24 commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,300 @@
+@[simp] lemma replace_row_self : replace_row A j b i j = b i := function.update_same j (b i) (A i)
+
+@[simp] lemma replace_column_eq {i' : n} (h : i = i') : replace_column A i b i' = b :=
+by {rw [h], apply replace_column_self}
```
New lines between lemmas.

ChrisHughes24 commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,300 @@
+  rw [pi.smul_apply, smul_eq_mul, mul_ite]
+end
+
+lemma mul_adjugate (A : matrix n n α) : A ⬝ adjugate A = A.det • 1 := begin
```
new line for `begin`

Vierkantor committed:
whitespace

Insert space between `finset.filter` and the filter condition.

Co-Authored-By: sgouezel <sebastien.gouezel@univ-rennes1.fr>

Vierkantor committed:
Improve docstrings

Vierkantor committed:
Make argument to `prod_cancels_of_partition_cancels` explicit

Vierkantor committed:
Rename `replace_column` and `replace_row` to `update_column` and `update_row`

Vierkantor committed:
Replace `update_column_eq` with `update_column_self` + rewriting step

Vierkantor committed:
whitespace

Newlines between all lemmas

Vierkantor committed:
whitespace

Newline before 'begin'

Vierkantor committed:
Fix conflicts with latest mathlib

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -0,0 +1,322 @@
+
+/-- Applying Cramer's rule to a column of the matrix gives a scaled basis vector. -/
+lemma cramer_column_self (i : n) :
+(@cramer _ _ _ α _ A) (A i) = (λ j, if i = j then A.det else 0) :=
```
Is the `@` still necessary here?

Vierkantor replied:
It isn't, good point!

Vierkantor committed:
Remove unnecessary explicitification of arguments

mergify[bot] committed:
Merge branch 'master' into matrix_inverse
